// @dendronhq/engine-server/lib/workspace/service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceService = exports.SyncActionStatus = void 0;
const common_all_1 = require("@dendronhq/common-all");
const common_server_1 = require("@dendronhq/common-server");
const fs_extra_1 = __importDefault(require("fs-extra"));
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
const config_1 = require("../config");
const metadata_1 = require("../metadata");
const migrations_1 = require("../migrations");
const seed_1 = require("../seed");
const git_1 = require("../topics/git");
const utils_1 = require("../utils");
const utils_2 = require("./utils");
const vscode_1 = require("./vscode");
const DENDRON_WS_NAME = common_all_1.CONSTANTS.DENDRON_WS_NAME;
var SyncActionStatus;
(function (SyncActionStatus) {
    SyncActionStatus["DONE"] = "";
    SyncActionStatus["NO_CHANGES"] = "it has no changes";
    SyncActionStatus["UNCOMMITTED_CHANGES"] = "it has uncommitted changes";
    SyncActionStatus["NO_REMOTE"] = "it has no remote";
    SyncActionStatus["NO_UPSTREAM"] = "the current branch has no upstream";
    SyncActionStatus["SKIP_CONFIG"] = "it is configured so";
    SyncActionStatus["NOT_PERMITTED"] = "user is not permitted to push to one or more vaults";
    SyncActionStatus["NEW"] = "newly clond repository";
    SyncActionStatus["ERROR"] = "error while syncing";
})(SyncActionStatus = exports.SyncActionStatus || (exports.SyncActionStatus = {}));
class WorkspaceService {
    constructor({ wsRoot, seedService }) {
        this.wsRoot = wsRoot;
        this.logger = common_server_1.createLogger();
        this._seedService = seedService || new seed_1.SeedService({ wsRoot });
    }
    static isNewVersionGreater({ oldVersion, newVersion, }) {
        return common_all_1.DUtils.semver.lt(oldVersion, newVersion);
    }
    static async isWorkspaceVault(fpath) {
        return fs_extra_1.default.pathExists(path_1.default.join(fpath, common_all_1.CONSTANTS.DENDRON_CONFIG_FILE));
    }
    get user() {
        const fpath = path_1.default.join(this.wsRoot, common_all_1.CONSTANTS.DENDRON_USER_FILE);
        if (fs_extra_1.default.existsSync(fpath)) {
            return new common_all_1.DUser(lodash_1.default.trim(fs_extra_1.default.readFileSync(fpath, { encoding: "utf8" })));
        }
        else {
            return common_all_1.DUser.createAnonymous();
        }
    }
    get config() {
        return config_1.DConfig.getOrCreate(this.wsRoot);
    }
    get dendronRoot() {
        return path_1.default.join(this.wsRoot, "dendron");
    }
    get seedService() {
        return this._seedService;
    }
    async setConfig(config) {
        const wsRoot = this.wsRoot;
        return config_1.DConfig.writeConfig({ wsRoot, config });
    }
    setWorkspaceConfig(config) {
        common_server_1.writeJSONWithComments(path_1.default.join(this.wsRoot, "dendron.code-workspace"), config);
    }
    /**
     *
     * @param param0
     * @returns `{vaults}` that have been added
     */
    async addWorkspace({ workspace }) {
        const config = this.config;
        const allWorkspaces = common_all_1.ConfigUtils.getWorkspace(config).workspaces || {};
        allWorkspaces[workspace.name] = lodash_1.default.omit(workspace, ["name", "vaults"]);
        // update vault
        const newVaults = await lodash_1.default.reduce(workspace.vaults, async (acc, vault) => {
            const out = await acc;
            out.push(await this.addVault({
                config,
                vault: { ...vault, workspace: workspace.name },
                updateConfig: false,
            }));
            return out;
        }, Promise.resolve([]));
        common_all_1.ConfigUtils.setWorkspaceProp(config, "workspaces", allWorkspaces);
        this.setConfig(config);
        return { vaults: newVaults };
    }
    /**
     *
     *
     * @param opts.vault - {@link DVault} to add to workspace
     * @param opts.config - if passed it, make modifications on passed in config instead of {wsRoot}/dendron.yml
     * @param opts.writeConfig - default: true, add to dendron.yml
     * @param opts.addToWorkspace - default: false, add to dendron.code-workspace. Make sure to keep false for Native workspaces.
     * @returns
     */
    async addVault(opts) {
        const { vault, config, updateConfig, updateWorkspace } = lodash_1.default.defaults(opts, {
            config: this.config,
            updateConfig: true,
            updateWorkspace: false,
        });
        const vaults = common_all_1.ConfigUtils.getVaults(config);
        vaults.unshift(vault);
        common_all_1.ConfigUtils.setVaults(config, vaults);
        // update dup note behavior
        if (!config.site.duplicateNoteBehavior) {
            const vaults = common_all_1.ConfigUtils.getVaults(config);
            config.site.duplicateNoteBehavior = {
                action: common_all_1.DuplicateNoteAction.USE_VAULT,
                payload: vaults.map((v) => common_all_1.VaultUtils.getName(v)),
            };
        }
        else if (lodash_1.default.isArray(config.site.duplicateNoteBehavior.payload)) {
            config.site.duplicateNoteBehavior.payload.push(common_all_1.VaultUtils.getName(vault));
        }
        if (updateConfig) {
            await this.setConfig(config);
        }
        if (updateWorkspace) {
            const wsPath = path_1.default.join(this.wsRoot, DENDRON_WS_NAME);
            let out = (await common_server_1.readJSONWithComments(wsPath));
            if (!lodash_1.default.find(out.folders, (ent) => ent.path === common_all_1.VaultUtils.getRelPath(vault))) {
                const vault2Folder = common_all_1.VaultUtils.toWorkspaceFolder(vault);
                const folders = [vault2Folder].concat(out.folders);
                out = common_server_1.assignJSONWithComment({ folders }, out);
                if (opts.onUpdatingWorkspace) {
                    await opts.onUpdatingWorkspace();
                }
                common_server_1.writeJSONWithComments(wsPath, out);
                if (opts.onUpdatedWorkspace) {
                    await opts.onUpdatedWorkspace();
                }
            }
        }
        else {
            // Run the hooks even if not updating the workspace file (native workspace), because other code depends on it.
            if (opts.onUpdatingWorkspace) {
                await opts.onUpdatingWorkspace();
            }
            if (opts.onUpdatedWorkspace) {
                await opts.onUpdatedWorkspace();
            }
        }
        return vault;
    }
    /**
     * Create vault files if it does not exist
     * @param opts.noAddToConfig: don't add to dendron.yml
     * @param opts.addToCodeWorkspace: add to dendron.code-workspace
     * @returns void
     *
     * Effects:
     *   - updates `dendron.yml` if `noAddToConfig` is not set
     *   - create directory
     *   - create root note and root schema
     */
    async createVault(opts) {
        const { vault, noAddToConfig } = opts;
        const vpath = common_server_1.vault2Path({ vault, wsRoot: this.wsRoot });
        fs_extra_1.default.ensureDirSync(vpath);
        const note = common_all_1.NoteUtils.createRoot({
            vault,
            body: [
                "# Welcome to Dendron",
                "",
                `This is the root of your dendron vault. If you decide to publish your entire vault, this will be your landing page. You are free to customize any part of this page except the frontmatter on top. `,
            ].join("\n"),
        });
        const schema = common_all_1.SchemaUtils.createRootModule({ vault });
        if (!fs_extra_1.default.existsSync(common_all_1.NoteUtils.getFullPath({ note, wsRoot: this.wsRoot }))) {
            await common_server_1.note2File({ note, vault, wsRoot: this.wsRoot });
        }
        if (!fs_extra_1.default.existsSync(common_all_1.SchemaUtils.getPath({ root: vpath, fname: "root" }))) {
            await common_server_1.schemaModuleOpts2File(schema, vpath, "root");
        }
        if (!noAddToConfig) {
            await this.addVault({ ...opts, updateWorkspace: false });
        }
        if (opts.addToCodeWorkspace || opts.updateWorkspace) {
            await this.addVaultToCodeWorkspace(vault);
        }
        return vault;
    }
    /** For vaults in the same repository, ensure that their sync configurations do not conflict. Returns the coordinated sync config. */
    verifyVaultSyncConfigs(vaults) {
        let prevVault;
        for (const vault of vaults) {
            if (lodash_1.default.isUndefined(vault.sync))
                continue;
            if (lodash_1.default.isUndefined(prevVault)) {
                prevVault = vault;
                continue;
            }
            if (prevVault.sync === vault.sync)
                continue;
            const prevVaultName = prevVault.name || prevVault.fsPath;
            const vaultName = vault.name || vault.fsPath;
            throw new common_all_1.DendronError({
                message: `Vaults ${prevVaultName} and ${vaultName} are in the same repository, but have conflicting configurations ${prevVault.sync} and ${vault.sync} set. Please remove conflicting configuration, or move vault to a different repository.`,
            });
        }
        return prevVault === null || prevVault === void 0 ? void 0 : prevVault.sync;
    }
    /** Checks if a given git command should be used on the vault based on user configuration.
     *
     * @param command The git command that we want to perform.
     * @param repo The location of the repository containing the vaults.
     * @param vaults The vaults on which the operation is being performed on.
     * @returns true if the command can be performed, false otherwise.
     */
    async shouldVaultsSync(command, [root, vaults]) {
        let workspaceVaultSyncConfig = this.verifyVaultSyncConfigs(vaults);
        if (lodash_1.default.isUndefined(workspaceVaultSyncConfig)) {
            if (await WorkspaceService.isWorkspaceVault(root)) {
                workspaceVaultSyncConfig = common_all_1.ConfigUtils.getWorkspace(this.config)
                    .workspaceVaultSyncMode;
                // default for workspace vaults
                if (lodash_1.default.isUndefined(workspaceVaultSyncConfig)) {
                    workspaceVaultSyncConfig = common_all_1.DVaultSync.NO_COMMIT;
                }
            }
            // default for regular vaults
            else
                workspaceVaultSyncConfig = common_all_1.DVaultSync.SYNC;
        }
        if (workspaceVaultSyncConfig === common_all_1.DVaultSync.SKIP)
            return false;
        if (workspaceVaultSyncConfig === common_all_1.DVaultSync.SYNC)
            return true;
        if (workspaceVaultSyncConfig === common_all_1.DVaultSync.NO_COMMIT &&
            command === "commit")
            return false;
        if (workspaceVaultSyncConfig === common_all_1.DVaultSync.NO_PUSH && command === "push")
            return false;
        return true;
    }
    async commitAndAddAll() {
        const allReposVaults = await this.getAllReposVaults();
        const out = await Promise.all(lodash_1.default.map([...allReposVaults.entries()], async (rootVaults) => {
            const [repo, vaults] = rootVaults;
            const git = new git_1.Git({ localUrl: repo });
            if (!(await this.shouldVaultsSync("commit", rootVaults)))
                return { repo, vaults, status: SyncActionStatus.SKIP_CONFIG };
            if (!(await git.hasChanges()))
                return { repo, vaults, status: SyncActionStatus.NO_CHANGES };
            try {
                await git.addAll();
                await git.commit({ msg: "update" });
                return { repo, vaults, status: SyncActionStatus.DONE };
            }
            catch (err) {
                const stderr = err.stderr ? `: ${err.stderr}` : "";
                throw new common_all_1.DendronError({
                    message: `error adding and committing vault${stderr}`,
                    payload: { err, repoPath: repo },
                });
            }
        }));
        return out;
    }
    /**
     * Initialize all remote vaults
     * @param opts
     * @returns
     */
    async initialize(opts) {
        const { onSyncVaultsProgress, onSyncVaultsEnd } = lodash_1.default.defaults(opts, {
            onSyncVaultsProgress: () => { },
            onSyncVaultsEnd: () => { },
        });
        const initializeRemoteVaults = common_all_1.ConfigUtils.getWorkspace(this.config).enableRemoteVaultInit;
        if (initializeRemoteVaults) {
            const { didClone } = await this.syncVaults({
                config: this.config,
                progressIndicator: onSyncVaultsProgress,
            });
            if (didClone) {
                onSyncVaultsEnd();
            }
            return didClone;
        }
        return false;
    }
    /**
     * Remove vaults. Currently doesn't delete any files.
     * @param param0
     */
    async removeVault(opts) {
        const config = this.config;
        const { vault, updateConfig, updateWorkspace } = lodash_1.default.defaults(opts, {
            updateConfig: true,
            updateWorkspace: false,
        });
        const vaults = common_all_1.ConfigUtils.getVaults(config);
        const vaultsAfterReject = lodash_1.default.reject(vaults, (ent) => {
            const checks = [
                common_all_1.VaultUtils.getRelPath(ent) === common_all_1.VaultUtils.getRelPath(vault),
            ];
            if (vault.workspace) {
                checks.push(ent.workspace === vault.workspace);
            }
            return lodash_1.default.every(checks);
        });
        common_all_1.ConfigUtils.setVaults(config, vaultsAfterReject);
        const workspaces = common_all_1.ConfigUtils.getWorkspace(config).workspaces;
        if (vault.workspace && workspaces) {
            const vaultWorkspace = lodash_1.default.find(common_all_1.ConfigUtils.getVaults(config), {
                workspace: vault.workspace,
            });
            if (lodash_1.default.isUndefined(vaultWorkspace)) {
                delete workspaces[vault.workspace];
                common_all_1.ConfigUtils.setWorkspaceProp(config, "workspaces", workspaces);
            }
        }
        if (config.site.duplicateNoteBehavior &&
            lodash_1.default.isArray(config.site.duplicateNoteBehavior.payload)) {
            const vaults = common_all_1.ConfigUtils.getVaults(config);
            if (vaults.length === 1) {
                // if there is only one vault left, remove duplicateNoteBehavior setting
                config.site = lodash_1.default.omit(config.site, ["duplicateNoteBehavior"]);
            }
            else {
                // otherwise pull the removed vault from payload
                config.site.duplicateNoteBehavior.payload = lodash_1.default.pull(config.site.duplicateNoteBehavior.payload, vault.fsPath);
            }
        }
        if (updateConfig) {
            await this.setConfig(config);
        }
        if (updateWorkspace) {
            const wsPath = path_1.default.join(this.wsRoot, DENDRON_WS_NAME);
            let settings = (await common_server_1.readJSONWithComments(wsPath));
            const folders = lodash_1.default.reject(settings.folders, (ent) => ent.path === common_all_1.VaultUtils.getRelPath(vault));
            settings = common_server_1.assignJSONWithComment({ folders }, settings);
            if (opts.onUpdatingWorkspace) {
                opts.onUpdatingWorkspace();
            }
            common_server_1.writeJSONWithComments(wsPath, settings);
            if (opts.onUpdatedWorkspace) {
                await opts.onUpdatedWorkspace();
            }
        }
        else {
            // Run the hooks even if not updating the workspace file (native workspace), because other code depends on it.
            if (opts.onUpdatingWorkspace) {
                opts.onUpdatingWorkspace();
            }
            if (opts.onUpdatedWorkspace) {
                await opts.onUpdatedWorkspace();
            }
        }
    }
    createConfig() {
        this.config;
    }
    /**
     * Initialize workspace with specified vaults
     * Files and folders created:
     * wsRoot/
     * - .gitignore
     * - dendron.yml
     * - {vaults}/
     *   - root.md
     *   - root.schema.yml
     *
     * NOTE: dendron.yml only gets created if you are adding a workspace...
     * @param opts
     */
    static async createWorkspace(opts) {
        const { wsRoot, vaults } = opts;
        const ws = new WorkspaceService({ wsRoot });
        fs_extra_1.default.ensureDirSync(wsRoot);
        // this creates `dendron.yml`
        ws.createConfig();
        // add gitignore
        const gitIgnore = path_1.default.join(wsRoot, ".gitignore");
        fs_extra_1.default.writeFileSync(gitIgnore, ["node_modules", ".dendron.*", "build", "\n"].join("\n"), { encoding: "utf8" });
        if (opts.createCodeWorkspace) {
            vscode_1.WorkspaceConfig.write(wsRoot, vaults);
        }
        await lodash_1.default.reduce(vaults, async (prev, vault) => {
            await prev;
            await ws.createVault({ vault });
            return;
        }, Promise.resolve());
        // check if this is the first workspace created
        if (lodash_1.default.isUndefined(metadata_1.MetadataService.instance().getMeta().firstWsInitialize)) {
            metadata_1.MetadataService.instance().setFirstWsInitialize();
        }
        return ws;
    }
    static async createFromConfig(opts) {
        const { wsRoot } = opts;
        const config = config_1.DConfig.getOrCreate(wsRoot);
        const ws = new WorkspaceService({ wsRoot });
        const vaults = common_all_1.ConfigUtils.getVaults(config);
        await Promise.all(vaults.map(async (vault) => {
            return ws.cloneVaultWithAccessToken({ vault });
        }));
        return;
    }
    async addVaultToCodeWorkspace(vault) {
        const wsRoot = this.wsRoot;
        // workspace file
        const wsPath = vscode_1.WorkspaceConfig.workspaceFile(wsRoot);
        let out = (await common_server_1.readJSONWithComments(wsPath));
        if (!lodash_1.default.find(out.folders, (ent) => ent.path === common_all_1.VaultUtils.getRelPath(vault))) {
            const vault2Folder = common_all_1.VaultUtils.toWorkspaceFolder(vault);
            const folders = [vault2Folder].concat(out.folders);
            out = common_server_1.assignJSONWithComment({ folders }, out);
            common_server_1.writeJSONWithComments(wsPath, out);
        }
        return;
    }
    /**
     * Used in createFromConfig
     */
    async cloneVaultWithAccessToken(opts) {
        const { vault } = opts;
        if (!vault.remote || vault.remote.type !== "git") {
            throw new common_all_1.DendronError({ message: "cloning non-git vault" });
        }
        let remotePath = vault.remote.url;
        const localPath = common_server_1.vault2Path({ vault, wsRoot: this.wsRoot });
        const git = common_server_1.simpleGit();
        this.logger.info({ msg: "cloning", remotePath, localPath });
        const accessToken = process.env["GITHUB_ACCESS_TOKEN"];
        if (accessToken) {
            this.logger.info({ msg: "using access token" });
            remotePath = common_server_1.GitUtils.getGithubAccessTokenUrl({
                remotePath,
                accessToken,
            });
        }
        await git.clone(remotePath, localPath);
    }
    /**
     * Clone a vault from a remote source
     * @param opts.vault vaults field
     * @param opts.urlTransformer modify the git url
     */
    async cloneVault(opts) {
        const { vault, urlTransformer } = lodash_1.default.defaults(opts, {
            urlTransformer: lodash_1.default.identity,
        });
        const wsRoot = this.wsRoot;
        if (!vault.remote || vault.remote.type !== "git") {
            throw new common_all_1.DendronError({ message: "cloning non-git vault" });
        }
        const repoPath = common_server_1.vault2Path({ wsRoot, vault });
        this.logger.info({ msg: "cloning", repoPath });
        const git = common_server_1.simpleGit({ baseDir: wsRoot });
        await git.clone(urlTransformer(vault.remote.url), repoPath);
        return repoPath;
    }
    async cloneWorkspace(opts) {
        const { wsRoot, urlTransformer, workspace, wsName } = lodash_1.default.defaults(opts, {
            urlTransformer: lodash_1.default.identity,
        });
        const repoPath = path_1.default.join(wsRoot, wsName);
        const git = common_server_1.simpleGit({ baseDir: wsRoot });
        await git.clone(urlTransformer(workspace.remote.url), wsName);
        return repoPath;
    }
    async getVaultRepo(vault) {
        const vpath = common_server_1.vault2Path({ vault, wsRoot: this.wsRoot });
        return await common_server_1.GitUtils.getGitRoot(vpath);
    }
    async getAllReposVaults() {
        const reposVaults = new Map();
        const vaults = common_all_1.ConfigUtils.getVaults(this.config);
        await Promise.all(vaults.map(async (vault) => {
            const repo = await this.getVaultRepo(vault);
            if (lodash_1.default.isUndefined(repo))
                return;
            const vaultsForRepo = reposVaults.get(repo) || [];
            vaultsForRepo.push(vault);
            reposVaults.set(repo, vaultsForRepo);
        }));
        return reposVaults;
    }
    async getAllRepos() {
        return [...(await this.getAllReposVaults()).keys()];
    }
    /**
     * Check if a path belongs to a workspace
     @deprecated - use {@link WorkspaceUtils.isPathInWorkspace}
     */
    isPathInWorkspace(fpath) {
        const vaults = common_all_1.ConfigUtils.getVaults(this.config);
        const wsRoot = this.wsRoot;
        return utils_2.WorkspaceUtils.isPathInWorkspace({ fpath, vaults, wsRoot });
    }
    async pullVault(opts) {
        const { vault } = lodash_1.default.defaults(opts, {
            urlTransformer: lodash_1.default.identity,
        });
        const wsRoot = this.wsRoot;
        if (!vault.remote || vault.remote.type !== "git") {
            throw new common_all_1.DendronError({ message: "pulling non-git vault" });
        }
        const repoPath = common_server_1.vault2Path({ wsRoot, vault });
        this.logger.info({ msg: "pulling ", repoPath });
        const git = common_server_1.simpleGit({ baseDir: repoPath });
        await git.pull();
        return repoPath;
    }
    /** Returns the list of vaults that were attempted to be pulled, even if there was nothing to pull. */
    async pullVaults() {
        const allReposVaults = await this.getAllReposVaults();
        const out = await Promise.all(lodash_1.default.map([...allReposVaults.entries()], async (rootVaults) => {
            const [repo, vaults] = rootVaults;
            const git = new git_1.Git({ localUrl: repo });
            // It's impossible to pull if there is no remote, or if there are tracked files that have changes
            if (!(await git.hasRemote()))
                return { repo, vaults, status: SyncActionStatus.NO_REMOTE };
            if (lodash_1.default.isUndefined(await git.getUpstream()))
                return { repo, vaults, status: SyncActionStatus.NO_UPSTREAM };
            if (!(await this.shouldVaultsSync("pull", rootVaults)))
                return { repo, vaults, status: SyncActionStatus.SKIP_CONFIG };
            if (await git.hasChanges({ untrackedFiles: "no" }))
                return {
                    repo,
                    vaults,
                    status: SyncActionStatus.UNCOMMITTED_CHANGES,
                };
            try {
                await git.pull();
                return { repo, vaults, status: SyncActionStatus.DONE };
            }
            catch (err) {
                const stderr = err.stderr ? `: ${err.stderr}` : "";
                throw new common_all_1.DendronError({
                    message: `error pulling vault${stderr}`,
                    payload: { err, repoPath: repo },
                });
            }
        }));
        return out;
    }
    /** Returns the list of vaults that were attempted to be pushed, even if there was nothing to push. */
    async pushVaults() {
        const allReposVaults = await this.getAllReposVaults();
        const out = await Promise.all(lodash_1.default.map([...allReposVaults.entries()], async (rootVaults) => {
            const [repo, vaults] = rootVaults;
            const git = new git_1.Git({ localUrl: repo });
            if (!(await git.hasRemote()))
                return { repo, vaults, status: SyncActionStatus.NO_REMOTE };
            const upstream = await git.getUpstream();
            if (lodash_1.default.isUndefined(upstream))
                return { repo, vaults, status: SyncActionStatus.NO_UPSTREAM };
            if ((await git.diff({
                nameOnly: true,
                oldCommit: upstream,
                newCommit: "HEAD",
            })) === "")
                return { repo, vaults, status: SyncActionStatus.NO_CHANGES };
            if (!(await this.shouldVaultsSync("push", rootVaults)))
                return { repo, vaults, status: SyncActionStatus.SKIP_CONFIG };
            if (!lodash_1.default.every(lodash_1.default.map(vaults, this.user.canPushVault)))
                return { repo, vaults, status: SyncActionStatus.NOT_PERMITTED };
            try {
                await git.push();
                return { repo, vaults, status: SyncActionStatus.DONE };
            }
            catch (err) {
                const stderr = err.stderr ? `: ${err.stderr}` : "";
                throw new common_all_1.DendronError({
                    message: `error pushing vault${stderr}`,
                    payload: { err, repoPath: repo },
                });
            }
        }));
        return out;
    }
    /**
     * Remove all vault caches in workspace
     */
    async removeVaultCaches() {
        const vaults = common_all_1.ConfigUtils.getVaults(this.config);
        await Promise.all(vaults.map((vault) => {
            return utils_1.removeCache(common_server_1.vault2Path({ wsRoot: this.wsRoot, vault }));
        }));
    }
    /**
     * See if there's anythign we need to change with the configuration
     */
    async runMigrationsIfNecessary({ forceUpgrade, workspaceInstallStatus, currentVersion, previousVersion, dendronConfig, wsConfig, }) {
        let changes = [];
        if (migrations_1.MigrationServce.shouldRunMigration({
            force: forceUpgrade,
            workspaceInstallStatus,
        })) {
            changes = await migrations_1.MigrationServce.applyMigrationRules({
                currentVersion,
                previousVersion,
                dendronConfig,
                wsConfig,
                wsService: this,
                logger: this.logger,
            });
            // if changes were made, use updated changes in subsequent configuration
            if (!lodash_1.default.isEmpty(changes)) {
                const { data } = lodash_1.default.last(changes);
                dendronConfig = data.dendronConfig;
            }
        }
        return changes;
    }
    async runConfigMigrationIfNecessary({ currentVersion, dendronConfig, }) {
        let changes = [];
        if (dendronConfig.version !== common_all_1.CURRENT_CONFIG_VERSION) {
            // we are on a legacy config.
            changes = await migrations_1.MigrationServce.applyMigrationRules({
                currentVersion,
                previousVersion: "0.64.1",
                dendronConfig,
                wsService: this,
                logger: this.logger,
                migrations: [migrations_1.CONFIG_MIGRATIONS],
            });
            // if changes were made, use updated changes in subsequent configuration
            if (!lodash_1.default.isEmpty(changes)) {
                const { data } = lodash_1.default.last(changes);
                dendronConfig = data.dendronConfig;
            }
        }
        return changes;
    }
    /**
     * Make sure all vaults are present on file system
     * @param fetchAndPull for repositories that exist, should we also do a fetch? default: false
     * @param skipPrivate skip cloning and pulling of private vaults. default: false
     */
    async syncVaults(opts) {
        const ctx = "syncVaults";
        const { config, progressIndicator, urlTransformer, fetchAndPull } = lodash_1.default.defaults(opts, { fetchAndPull: false, skipPrivate: false });
        const { wsRoot } = this;
        const workspaces = common_all_1.ConfigUtils.getWorkspace(config).workspaces;
        // check workspaces
        const workspacePaths = (await Promise.all(lodash_1.default.map(workspaces, async (wsEntry, wsName) => {
            const wsPath = path_1.default.join(wsRoot, wsName);
            if (!fs_extra_1.default.existsSync(wsPath)) {
                return {
                    wsPath: await this.cloneWorkspace({
                        wsName,
                        workspace: wsEntry,
                        wsRoot,
                    }),
                    wsUrl: wsEntry.remote.url,
                };
            }
            return;
        }))).filter((ent) => !lodash_1.default.isUndefined(ent));
        // const seedService = new SeedService({wsRoot});
        // check seeds
        const seeds = common_all_1.ConfigUtils.getWorkspace(config).seeds;
        const seedResults = [];
        await Promise.all(lodash_1.default.map(seeds, async (entry, id) => {
            if (!(await seed_1.SeedUtils.exists({ id, wsRoot }))) {
                const resp = await this._seedService.info({ id });
                if (lodash_1.default.isUndefined(resp)) {
                    seedResults.push({
                        id,
                        status: SyncActionStatus.ERROR,
                        data: new common_all_1.DendronError({
                            status: SyncActionStatus.ERROR,
                            message: `seed ${id} does not exist in registry`,
                        }),
                    });
                    return;
                }
                const spath = await this._seedService.cloneSeed({
                    seed: resp,
                    branch: entry.branch,
                });
                seedResults.push({
                    id,
                    status: SyncActionStatus.NEW,
                    data: { spath },
                });
            }
            return undefined;
        }));
        // clone all missing vaults
        const vaults = common_all_1.ConfigUtils.getVaults(config);
        const emptyRemoteVaults = vaults.filter((vault) => !lodash_1.default.isUndefined(vault.remote) &&
            !fs_extra_1.default.existsSync(common_server_1.vault2Path({ vault, wsRoot })));
        const didClone = !lodash_1.default.isEmpty(emptyRemoteVaults) ||
            !lodash_1.default.isEmpty(workspacePaths) ||
            !lodash_1.default.isUndefined(seedResults.find((ent) => ent.status === SyncActionStatus.NEW));
        // if we added a workspace, we also add new vaults
        if (!lodash_1.default.isEmpty(workspacePaths)) {
            this.setConfig(config);
        }
        if (progressIndicator && didClone) {
            progressIndicator();
        }
        await Promise.all(emptyRemoteVaults.map(async (vault) => {
            return this.cloneVault({ vault, urlTransformer });
        }));
        if (fetchAndPull) {
            const vaults = common_all_1.ConfigUtils.getVaults(config);
            const vaultsToFetch = lodash_1.default.difference(vaults.filter((vault) => !lodash_1.default.isUndefined(vault.remote)), emptyRemoteVaults);
            this.logger.info({ ctx, msg: "fetching vaults", vaultsToFetch });
            await Promise.all(vaultsToFetch.map(async (vault) => {
                return this.pullVault({ vault });
            }));
        }
        return { didClone };
    }
    writePort(port) {
        const wsRoot = this.wsRoot;
        // dendron-cli can overwrite port file. anything that needs the port should connect to `portFilePathExtension`
        const portFilePath = utils_1.getPortFilePath({ wsRoot });
        fs_extra_1.default.writeFileSync(portFilePath, lodash_1.default.toString(port), { encoding: "utf8" });
    }
    writeMeta(opts) {
        const { version } = opts;
        const fpath = utils_1.getWSMetaFilePath({ wsRoot: this.wsRoot });
        return utils_1.writeWSMetaFile({
            fpath,
            data: {
                version,
                activationTime: common_all_1.Time.now().toMillis(),
            },
        });
    }
}
exports.WorkspaceService = WorkspaceService;
