// @fec/eleventy-plugin-remark/dist/eleventy.cjs.js
'use strict';

var remark = require('remark');
var remarkRehype = require('remark-rehype');
var rehypeStringify = require('rehype-stringify');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var remark__default = /*#__PURE__*/_interopDefaultLegacy(remark);
var remarkRehype__default = /*#__PURE__*/_interopDefaultLegacy(remarkRehype);
var rehypeStringify__default = /*#__PURE__*/_interopDefaultLegacy(rehypeStringify);

function requirePlugins(plugins) {
  if (!Array.isArray(plugins)) {
    throw new Error('plugins option is not an array');
  }

  const requirePlugin = (item) => {
    if (typeof item === 'function') {
      return item;
    } else if (typeof item === 'string') {
      return require(item);
    }

    throw new Error(
      `plugin has to be a function or a string, ${typeof item} type passed`
    );
  };

  const list = plugins.map((item) => {
    let options = {};
    let fn;
    if (typeof item === 'object' && item !== null && item.plugin) {
      fn = requirePlugin(item.plugin);
      options = item.options ? item.options : {};
    } else {
      fn = requirePlugin(item);
    }

    return [fn, options];
  });

  return list;
}

function eleventyRemark(options) {
  const processor = remark__default["default"]();
  const plugins = requirePlugins(options.plugins);
  processor.use(plugins);

  if (options.enableRehype) {
    processor.use(remarkRehype__default["default"]).use(rehypeStringify__default["default"]);
  }

  return {
    set: () => {},
    render: async (str) => {
      const { contents } = await processor.process(str);
      return contents;
    },
  };
}

const defaultEleventyRemarkOptions = {
  plugins: [],
  enableRehype: true,
};

var _eleventy = {
  initArguments: {},
  configFunction: (eleventyConfig, pluginOptions = {}) => {
    const options = Object.assign(
      {},
      defaultEleventyRemarkOptions,
      pluginOptions
    );
    eleventyConfig.setLibrary('md', eleventyRemark(options));
  },
};

module.exports = _eleventy;
