// @dendronhq/engine-server/lib/drivers/file/schemaParser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaParser = void 0;
const common_all_1 = require("@dendronhq/common-all");
const common_server_1 = require("@dendronhq/common-server");
const fs_extra_1 = __importDefault(require("fs-extra"));
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
const yamljs_1 = __importDefault(require("yamljs"));
const parseBase_1 = require("./parseBase");
class SchemaParser extends parseBase_1.ParserBase {
    async parseFile(fpath, root) {
        const fname = path_1.default.basename(fpath, ".schema.yml");
        const wsRoot = this.opts.store.wsRoot;
        const vpath = common_server_1.vault2Path({ vault: root, wsRoot });
        const schemaOpts = yamljs_1.default.parse(fs_extra_1.default.readFileSync(path_1.default.join(vpath, fpath), "utf8"));
        return await common_server_1.SchemaParserV2.parseRaw(schemaOpts, { root, fname, wsRoot });
    }
    async parse(fpaths, vault) {
        const ctx = "parse";
        this.logger.info({ ctx, msg: "enter", fpaths, vault });
        const out = await Promise.all(fpaths.flatMap(async (fpath) => {
            try {
                return await this.parseFile(fpath, vault);
            }
            catch (err) {
                return new common_all_1.DendronError({
                    message: common_all_1.ERROR_STATUS.BAD_PARSE_FOR_SCHEMA,
                    payload: { fpath },
                });
            }
        }));
        let errors = lodash_1.default.filter(out, (ent) => ent instanceof common_all_1.DendronError);
        return {
            schemas: lodash_1.default.reject(out, (ent) => ent instanceof common_all_1.DendronError),
            errors: lodash_1.default.isEmpty(errors) ? null : errors,
        };
    }
}
exports.SchemaParser = SchemaParser;
