// @dendronhq/common-all/lib/constants/configs/global.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLOBAL = exports.ENABLE_NOTE_TITLE_FOR_LINK = exports.HIERARCHY_DISPLAY_TITLE = exports.ENABLE_HIERARCHY_DISPLAY = exports.ENABLE_FM_TITLE = exports.ENABLE_KATEX = exports.ENABLE_PRETTY_REFS = exports.ENABLE_NUNJUCKS = exports.ENABLE_MERMAID = void 0;
const ENABLE_MERMAID = (namespace) => {
    return {
        label: `Enable Mermaid (${namespace})`,
        desc: `Enable the use of mermaid for rendering diagrams. (${namespace})`,
    };
};
exports.ENABLE_MERMAID = ENABLE_MERMAID;
const ENABLE_NUNJUCKS = (namespace) => {
    return {
        label: `Enable Nunjucks (${namespace})`,
        desc: `Enable the use of nunjucks templates in the note body. (${namespace})`,
    };
};
exports.ENABLE_NUNJUCKS = ENABLE_NUNJUCKS;
const ENABLE_PRETTY_REFS = (namespace) => {
    return {
        label: `Enable Pretty Refs (${namespace})`,
        desc: `Enable rendering note references as pretty refs. (${namespace})`,
    };
};
exports.ENABLE_PRETTY_REFS = ENABLE_PRETTY_REFS;
const ENABLE_KATEX = (namespace) => {
    return {
        label: `Enable Katex (${namespace})`,
        desc: `Enable the use of katex for rendering math. (${namespace})`,
    };
};
exports.ENABLE_KATEX = ENABLE_KATEX;
const ENABLE_FM_TITLE = (namespace) => {
    return {
        label: `Enable Frontmatter Title (${namespace})`,
        desc: `Insert frontmatter title of note to the body (${namespace})`,
    };
};
exports.ENABLE_FM_TITLE = ENABLE_FM_TITLE;
const ENABLE_HIERARCHY_DISPLAY = (namespace) => {
    return {
        label: `Enable Hierarchy Display (${namespace})`,
        desc: `Enable rendering of children link block at the end of the note. (${namespace})`,
    };
};
exports.ENABLE_HIERARCHY_DISPLAY = ENABLE_HIERARCHY_DISPLAY;
const HIERARCHY_DISPLAY_TITLE = (namespace) => {
    return {
        label: `Hierarhcy Display Title ${namespace}`,
        desc: `Title to display for the children links block. ${namespace}`,
    };
};
exports.HIERARCHY_DISPLAY_TITLE = HIERARCHY_DISPLAY_TITLE;
const ENABLE_NOTE_TITLE_FOR_LINK = (namespace) => {
    return {
        label: `Enable Note Title for Links (${namespace})`,
        desc: `Enable rendering of naked links as the title of the note. (${namespace})`,
    };
};
exports.ENABLE_NOTE_TITLE_FOR_LINK = ENABLE_NOTE_TITLE_FOR_LINK;
exports.GLOBAL = {
    enableFMTitle: exports.ENABLE_FM_TITLE("global"),
    enableHierarchyDisplay: exports.ENABLE_HIERARCHY_DISPLAY("global"),
    hierarchyDisplayTitle: exports.HIERARCHY_DISPLAY_TITLE("global"),
    enableNoteTitleForLink: exports.ENABLE_NOTE_TITLE_FOR_LINK("global"),
    enableMermaid: exports.ENABLE_MERMAID("global"),
    enableNunjucks: exports.ENABLE_NUNJUCKS("global"),
    enablePrettyRefs: exports.ENABLE_PRETTY_REFS("global"),
    enableKatex: exports.ENABLE_KATEX("global"),
};
