// @dendronhq/engine-server/lib/markdown/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VaultMissingBehavior = exports.DendronASTDest = exports.DendronASTTypes = void 0;
var DendronASTTypes;
(function (DendronASTTypes) {
    DendronASTTypes["WIKI_LINK"] = "wikiLink";
    DendronASTTypes["REF_LINK"] = "refLink";
    DendronASTTypes["REF_LINK_V2"] = "refLinkV2";
    DendronASTTypes["BLOCK_ANCHOR"] = "blockAnchor";
    DendronASTTypes["HASHTAG"] = "hashtag";
    DendronASTTypes["USERTAG"] = "usertag";
    DendronASTTypes["EXTENDED_IMAGE"] = "extendedImage";
    // Not dendron-specific, included here for convenience
    DendronASTTypes["ROOT"] = "root";
    DendronASTTypes["HEADING"] = "heading";
    DendronASTTypes["LIST"] = "list";
    DendronASTTypes["LIST_ITEM"] = "listItem";
    DendronASTTypes["PARAGRAPH"] = "paragraph";
    DendronASTTypes["TEXT"] = "text";
    DendronASTTypes["TABLE"] = "table";
    DendronASTTypes["TABLE_ROW"] = "tableRow";
    DendronASTTypes["TABLE_CELL"] = "tableCell";
    DendronASTTypes["IMAGE"] = "image";
    DendronASTTypes["FRONTMATTER"] = "yaml";
    DendronASTTypes["LINK"] = "link";
    DendronASTTypes["FOOTNOTE_DEFINITION"] = "footnoteDefinition";
    DendronASTTypes["FOOTNOTE_REFERENCE"] = "footnoteReference";
})(DendronASTTypes = exports.DendronASTTypes || (exports.DendronASTTypes = {}));
var DendronASTDest;
(function (DendronASTDest) {
    DendronASTDest["MD_ENHANCED_PREVIEW"] = "MD_ENHANCED_PREVIEW";
    DendronASTDest["MD_REGULAR"] = "MD_REGULAR";
    DendronASTDest["MD_DENDRON"] = "MD_DENDRON";
    DendronASTDest["HTML"] = "HTML";
})(DendronASTDest = exports.DendronASTDest || (exports.DendronASTDest = {}));
var VaultMissingBehavior;
(function (VaultMissingBehavior) {
    VaultMissingBehavior[VaultMissingBehavior["FALLBACK_TO_ORIGINAL_VAULT"] = 0] = "FALLBACK_TO_ORIGINAL_VAULT";
    VaultMissingBehavior[VaultMissingBehavior["THROW_ERROR"] = 1] = "THROW_ERROR";
})(VaultMissingBehavior = exports.VaultMissingBehavior || (exports.VaultMissingBehavior = {}));
