// @dendronhq/common-all/lib/constants/views.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWebEditorViewEntry = exports.getWebTreeViewEntry = exports.isWebViewEntry = exports.TREE_VIEWS = exports.EDITOR_VIEWS = exports.DendronTreeViewKey = exports.DendronEditorViewKey = void 0;
const __1 = require("..");
var DendronEditorViewKey;
(function (DendronEditorViewKey) {
    DendronEditorViewKey["CONFIGURE"] = "dendron.configure";
    DendronEditorViewKey["NOTE_GRAPH"] = "dendron.graph-note";
    DendronEditorViewKey["SCHEMA_GRAPH"] = "dendron.graph-schema";
    DendronEditorViewKey["NOTE_PREVIEW"] = "dendron.note-preview";
    DendronEditorViewKey["SEED_BROWSER"] = "dendron.seed-browser";
})(DendronEditorViewKey = exports.DendronEditorViewKey || (exports.DendronEditorViewKey = {}));
var DendronTreeViewKey;
(function (DendronTreeViewKey) {
    DendronTreeViewKey["SAMPLE_VIEW"] = "dendron.sample";
    DendronTreeViewKey["TREE_VIEW"] = "dendron.treeView";
    DendronTreeViewKey["TREE_VIEW_V2"] = "dendron.tree-view";
    DendronTreeViewKey["BACKLINKS"] = "dendron.backlinks";
    DendronTreeViewKey["CALENDAR_VIEW"] = "dendron.calendar-view";
    DendronTreeViewKey["LOOKUP_VIEW"] = "dendron.lookup-view";
})(DendronTreeViewKey = exports.DendronTreeViewKey || (exports.DendronTreeViewKey = {}));
exports.EDITOR_VIEWS = {
    [DendronEditorViewKey.NOTE_PREVIEW]: {
        desc: "Note Preview",
        label: "Note Preview",
        bundleName: "notePreview",
        type: "webview",
    },
    [DendronEditorViewKey.CONFIGURE]: {
        desc: "TODO",
        label: "TODO",
        type: "nativeview",
    },
    [DendronEditorViewKey.NOTE_GRAPH]: {
        desc: "TODO",
        label: "TODO",
        type: "nativeview",
    },
    [DendronEditorViewKey.SCHEMA_GRAPH]: {
        desc: "TODO",
        label: "TODO",
        type: "nativeview",
    },
    [DendronEditorViewKey.SEED_BROWSER]: {
        desc: "TODO",
        label: "TODO",
        type: "nativeview",
    },
};
/**
 * Value is the name of webpack bundle for webview based tree views
 */
exports.TREE_VIEWS = {
    [DendronTreeViewKey.TREE_VIEW_V2]: {
        desc: "Tree View",
        label: "Tree View",
        bundleName: "treePanelView",
        type: "webview",
    },
    [DendronTreeViewKey.SAMPLE_VIEW]: {
        desc: "A view used for prototyping",
        label: "Sample View",
        bundleName: "TODO",
        type: "webview",
    },
    [DendronTreeViewKey.TREE_VIEW]: {
        desc: "Tree View",
        label: "Tree View",
        type: "nativeview",
    },
    [DendronTreeViewKey.BACKLINKS]: {
        desc: "Shows all backlinks to the currentnote",
        label: "Backlinks",
        type: "nativeview",
    },
    [DendronTreeViewKey.CALENDAR_VIEW]: {
        desc: "Calendar View",
        label: "Calendar View",
        type: "webview",
        bundleName: "TODO",
    },
    [DendronTreeViewKey.LOOKUP_VIEW]: {
        desc: "Lookup View",
        label: "Lookup View",
        type: "webview",
        bundleName: "LookupPanelView",
    },
};
const isWebViewEntry = (entry) => {
    return entry.type === "webview";
};
exports.isWebViewEntry = isWebViewEntry;
const getWebTreeViewEntry = (key) => {
    const out = exports.TREE_VIEWS[key];
    if ((0, exports.isWebViewEntry)(out)) {
        return out;
    }
    throw __1.ErrorFactory.createInvalidStateError({
        message: `${key} is not valid webview key`,
    });
};
exports.getWebTreeViewEntry = getWebTreeViewEntry;
const getWebEditorViewEntry = (key) => {
    const out = exports.EDITOR_VIEWS[key];
    if ((0, exports.isWebViewEntry)(out)) {
        return out;
    }
    throw __1.ErrorFactory.createInvalidStateError({
        message: `${key} is not valid webview key`,
    });
};
exports.getWebEditorViewEntry = getWebEditorViewEntry;
