// @dendronhq/common-all/lib/types/configs/publishing/publishing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genDefaultPublishingConfig = exports.DuplicateNoteActionEnum = void 0;
const github_1 = require("./github");
const seo_1 = require("./seo");
var DuplicateNoteActionEnum;
(function (DuplicateNoteActionEnum) {
    DuplicateNoteActionEnum["useVault"] = "useVault";
})(DuplicateNoteActionEnum = exports.DuplicateNoteActionEnum || (exports.DuplicateNoteActionEnum = {}));
/**
 * Generate default {@link DendronPublishingConfig}
 * @returns DendronPublishingConfig
 */
function genDefaultPublishingConfig() {
    return {
        copyAssets: true,
        siteHierarchies: ["root"],
        writeStubs: false,
        enableContainers: false,
        generateChangelog: false,
        siteRootDir: "docs",
        seo: seo_1.genDefaultSEOConfig(),
        github: github_1.genDefaultGithubConfig(),
        enableSiteLastModified: true,
        enableFrontmatterTags: true,
        enableRandomlyColoredTags: true,
        enablePrettyLinks: true,
    };
}
exports.genDefaultPublishingConfig = genDefaultPublishingConfig;
