// @dendronhq/common-all/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wrap = exports.normalizeUnixPath = exports.cleanName = exports.ConfigUtils = exports.PublishUtils = exports.TagUtils = exports.debounceAsyncUntilComplete = exports.throttleAsyncUntilComplete = exports.mapValues = exports.groupBy = exports.FIFOQueue = exports.NoteFNamesDict = exports.ListMap = exports.DefaultMap = exports.makeColorTranslucent = exports.randomColor = exports.isFalsy = exports.isNotUndefined = exports.isLineAnchor = exports.isBlockAnchor = exports.isNumeric = exports.getSlugger = exports.DUtils = void 0;
const github_slugger_1 = __importDefault(require("github-slugger"));
const lodash_1 = __importDefault(require("lodash"));
const minimatch_1 = __importDefault(require("minimatch"));
const path_1 = __importDefault(require("path"));
const querystring_1 = __importDefault(require("querystring"));
const semver_1 = __importDefault(require("semver"));
const constants_1 = require("./constants");
const error_1 = require("./error");
const colors_1 = require("./colors");
const compat_1 = require("./constants/configs/compat");
const intermediateConfigs_1 = require("./types/intermediateConfigs");
const regex_1 = require("./util/regex");
/**
 * Dendron utilities
 */
class DUtils {
}
exports.DUtils = DUtils;
DUtils.minimatch = minimatch_1.default;
DUtils.semver = semver_1.default;
DUtils.querystring = querystring_1.default;
const getSlugger = () => {
    return new github_slugger_1.default();
};
exports.getSlugger = getSlugger;
/**
 * determine if given parameter is numeric
 * https://stackoverflow.com/questions/18082/validate-decimal-numbers-in-javascript-isnumeric/1830844#1830844
 * @param n
 * @returns boolean
 */
const isNumeric = (n) => {
    // eslint-disable-next-line no-restricted-globals, radix
    return !isNaN(parseInt(n)) && isFinite(n);
};
exports.isNumeric = isNumeric;
function isBlockAnchor(anchor) {
    // not undefined, not an empty string, and the first character is ^
    return !!anchor && anchor[0] === "^";
}
exports.isBlockAnchor = isBlockAnchor;
function isLineAnchor(anchor) {
    // not undefined, not an empty string, and the first character is L, and is followed by numbers
    return !!anchor && /L\d+/.test(anchor);
}
exports.isLineAnchor = isLineAnchor;
/** A type guard for things that are not undefined.
 *
 * This is equivalent to !_.isUndefined(), except that it provides a type guard
 * ensuring the parameter is not undefined. This is useful when filtering:
 *
 * function foo(list: (string | undefined)[]) {
 *   const stringsOnly = list.filter(isNotUndefined);
 * }
 *
 * This will give stringsOnly the type string[]. Without the type guard, it would have
 * received the type (string | undefined)[] despite the fact that we filtered out undefined.
 */
function isNotUndefined(t) {
    return !lodash_1.default.isUndefined(t);
}
exports.isNotUndefined = isNotUndefined;
/**
 * Check if the value u is a falsy value.
 */
function isFalsy(u) {
    if (lodash_1.default.isBoolean(u)) {
        return u === false;
    }
    return lodash_1.default.some([lodash_1.default.isUndefined(u), lodash_1.default.isEmpty(u), lodash_1.default.isNull(u)]);
}
exports.isFalsy = isFalsy;
/** Calculates a basic integer hash for the given string.
 *
 * This is very unsafe, do not rely on this for anything where collisions are bad.
 *
 * Adapted from https://github.com/darkskyapp/string-hash.
 * Originally released under CC0 1.0 Universal (CC0 1.0) Public Domain Dedication.
 */
function basicStringHash(text) {
    // eslint-disable-next-line no-bitwise
    return (
    // eslint-disable-next-line no-bitwise
    lodash_1.default.reduce(text, (prev, curr) => {
        return prev + curr.charCodeAt(0);
    }, 5381) >>>
        // JavaScript does bitwise operations (like XOR, above) on 32-bit signed
        // integers. Since we want the results to be always positive, convert the
        // signed int to an unsigned by doing an unsigned bitshift.
        0);
}
/** Given a string, return a random color as a HTML color code.
 *
 * The same string will always generate the same color, and different strings will get different random colors.
 */
function randomColor(text) {
    return colors_1.COLORS_LIST[basicStringHash(text) % colors_1.COLORS_LIST.length];
}
exports.randomColor = randomColor;
/** Only some colors are recognized, other colors will be returned without being modified.
 *
 * Examples of recognized colors:
 * * #45AB35
 * * rgb(123, 23, 45)
 * * rgb(123 23 45)
 * * hsl(123, 23%, 45%)
 * * hsl(123 23% 45%)
 *
 * This function does not verify that the input colors are valid, but as long as a valid color is passed in
 * it will not generate an invalid color.
 *
 * @param color
 * @param translucency A number between 0 and 1, with 0 being fully transparent and 1 being fully opaque.
 * @returns
 */
function makeColorTranslucent(color, translucency) {
    let match = color.match(/^#[\dA-Fa-f]{6}$/);
    if (match)
        return `${color}${(translucency * 255).toString(16)}`;
    match = color.match(/^((rgb|hsl)\( *[\d.]+ *, *[\d.]+%? *, *[\d.]+%? *)\)$/);
    if (match)
        return `${match[1]}, ${translucency})`;
    match = color.match(/^((rgb|hsl)\( *[\d.]+ *[\d.]+%? *[\d.]+%? *)\)$/);
    if (match)
        return `${match[1]} / ${translucency})`;
    return color;
}
exports.makeColorTranslucent = makeColorTranslucent;
/** A map that automatically inserts a value provided by the factory when a missing key is looked up.
 *
 * Modeled after python's `defaultdict`.
 *
 * Mind that `get` may mutate the map, which may be unintuitive.
 *
 * Example usage:
 *
 * ```ts
 * const myMap = new DefaultMap<string, string[]>(() => []);
 * myMap.get("foo").push("bar");
 * ```
 */
class DefaultMap {
    constructor(factory) {
        this._internalMap = new Map();
        this._factory = factory;
    }
    get(key) {
        let value = this._internalMap.get(key);
        if (lodash_1.default.isUndefined(value)) {
            value = this._factory(key);
            this._internalMap.set(key, value);
        }
        return value;
    }
    set(key, value) {
        return this._internalMap.set(key, value);
    }
    has(key) {
        return this._internalMap.has(key);
    }
    keys() {
        return this._internalMap.keys();
    }
    values() {
        return this._internalMap.values();
    }
    entries() {
        return this._internalMap.entries();
    }
    delete(key) {
        return this._internalMap.delete(key);
    }
    get size() {
        return this._internalMap.size;
    }
}
exports.DefaultMap = DefaultMap;
/** Maps a `K` to a list of `V`s. */
class ListMap {
    constructor() {
        this._internalMap = new Map();
    }
    get(key) {
        return this._internalMap.get(key);
    }
    add(key, ...toAdd) {
        let values = this._internalMap.get(key);
        if (values === undefined)
            values = [];
        values.push(...toAdd);
        this._internalMap.set(key, values);
    }
    delete(key, ...toDelete) {
        const values = this._internalMap.get(key);
        if (values === undefined)
            return;
        lodash_1.default.pull(values, ...toDelete);
        if (values.length === 0) {
            this._internalMap.delete(key);
        }
        else {
            this._internalMap.set(key, values);
        }
    }
    has(key, value) {
        const values = this._internalMap.get(key);
        if (values === undefined)
            return false;
        return values.includes(value);
    }
}
exports.ListMap = ListMap;
class NoteFNamesDict {
    constructor(initialNotes) {
        this._internalMap = new ListMap();
        if (initialNotes)
            this.addAll(initialNotes);
    }
    get(notes, fname) {
        const keys = this._internalMap.get(cleanName(fname));
        if (keys === undefined)
            return [];
        return keys.map((key) => notes[key]).filter(isNotUndefined);
    }
    /** Returns true if dict has `note` exactly with this fname and id. */
    has(note) {
        var _a;
        return !!((_a = this._internalMap
            // there are notes with this fname
            .get(cleanName(note.fname))) === null || _a === void 0 ? void 0 : _a.some((maybeMatch) => maybeMatch === note.id));
    }
    add(note) {
        if (this.has(note))
            return; // avoid duplicates
        this._internalMap.add(cleanName(note.fname), note.id);
    }
    addAll(notes) {
        notes.forEach((note) => this.add(note));
    }
    delete(note) {
        this._internalMap.delete(cleanName(note.fname), note.id);
    }
}
exports.NoteFNamesDict = NoteFNamesDict;
class FIFOQueue {
    constructor(init) {
        this._internalQueue = [];
        if (init)
            this._internalQueue = init;
    }
    enqueue(item) {
        this._internalQueue.push(item);
    }
    enqueueAll(items) {
        for (const item of items)
            this.enqueue(item);
    }
    dequeue() {
        return this._internalQueue.shift();
    }
    get length() {
        return this._internalQueue.length;
    }
}
exports.FIFOQueue = FIFOQueue;
/** Similar to lodash `_.groupBy`, except not limited to string keys. */
function groupBy(collection, iteratee) {
    const map = new Map();
    collection.forEach((value, index) => {
        const key = iteratee(value, index);
        let group = map.get(key);
        if (group === undefined) {
            group = [];
            map.set(key, group);
        }
        group.push(value);
    });
    return map;
}
exports.groupBy = groupBy;
function mapValues(inMap, applyFn) {
    const outMap = new Map();
    for (const [key, value] of inMap.entries()) {
        outMap.set(key, applyFn(value));
    }
    return outMap;
}
exports.mapValues = mapValues;
/** Throttles a given async function so that it is only executed again once the first execution is complete.
 *
 * Similar to lodash _.throttle, except that:
 * 1. It's aware of the inputs, and will only throttle calls where `keyFn` returns the same key for the inputs of that call.
 * 2. Rather than a set timeout, it will keep throttling until the first async call is complete or if set, the `maxTimeout` is reached.
 *
 * @param fn The function to throttle.
 * @param keyFn A function that takes the inputs to an `fn` call and turns them into an identifying key, where to calls with same input will have the same key.
 * @param timeout Optional, in ms. If set, the throttle will not throttle for more than this much time. Once the timeout is reached, the next call will be allowed to execute.
 * @returns The throttled function. This function will return its results if it got executed, or undefined it it was throttled.
 */
function throttleAsyncUntilComplete({ fn, keyFn, timeout, }) {
    const lastStarted = new Map();
    return async (...args) => {
        const key = keyFn(...args);
        const last = lastStarted.get(key);
        if (last === undefined ||
            (timeout !== undefined && Date.now() - last > timeout)) {
            // Function was never run with this input before or it timed out, re-run it
            lastStarted.set(key, Date.now());
            let out;
            try {
                out = await fn(...args);
            }
            finally {
                lastStarted.delete(key);
            }
            return out;
        }
        return undefined;
    };
}
exports.throttleAsyncUntilComplete = throttleAsyncUntilComplete;
/** Debounces a given async function so that it is only executed again once the first execution is complete.
 *
 * Similar to lodash _.debounce, except that:
 * 1. It's aware of the inputs, and will only debounce calls where `keyFn` returns the same key for the inputs of that call.
 * 2. In addition to the timeout, it will also debouce calls while the async function is executing.
 *
 * Differently from `throttleAsyncUntilComplete`, this will wait for the timeout to expire before running the function for the first time.
 * Additionally, if any calls occur while the function is being executed and `trailing` is set,
 * another timeout and execution will happen once the current execution is done.
 * For example, consider this timeline where the arrows are calls to the debounced function.
 *
 * ```
 * +---------+---------+---------+---------+
 * | timeout | execute | timeout | execute |
 * +---------+---------+---------+---------+
 * ^   ^   ^     ^
 * ```
 * The timeout starts at first function call, all calls during that time are debounced, and the function finally executes after the timeout.
 * Because another function call happens during the execution of the async function, another timeout and execution trigger right after the
 * first is done (this will only happen if trailing is set). After that point, no more function calls occur so no more timeouts or executions happen.
 *
 * **Check `windowDecorations.ts` for an example of how this is used. It was primarily purpose-built for that.
 *
 * @param fn The function to debounce.
 * @param keyFn A function that takes the inputs to an `fn` call and turns them into an identifying key, where to calls with same input will have the same key.
 * @param timeout In ms. The function will not execute until this much time has passed. In other words, there will be at least this much time between executions.
 * @param trailing Optional. If set, an additional execution will be done to respond to calls during the execute phase.
 * @returns An object containing the debounced function, and the
 */
function debounceAsyncUntilComplete({ fn, keyFn, timeout, trailing, }) {
    const states = new Map();
    const debouncedFn = async (...args) => {
        const key = keyFn(...args);
        const state = states.get(key);
        if (state === "timeout" || state === "trailing") {
            // Another execution is already scheduled
            return;
        }
        else if (state === "execute" && trailing) {
            // Currently executing, and configured for a trailing execution
            states.set(key, "trailing");
        }
        else {
            // Not currently executing or scheduled, schedule now
            states.set(key, "timeout");
            setTimeout(async () => {
                // timeout done, start executing
                states.set(key, "execute");
                try {
                    await fn(...args);
                }
                finally {
                    const lastState = states.get(key);
                    // execution complete, mark as not executing
                    states.delete(key);
                    if (lastState === "trailing") {
                        // but if we had a trailing execution scheduled, do that
                        debouncedFn(...args);
                    }
                }
            }, timeout);
        }
    };
    return { debouncedFn, states };
}
exports.debounceAsyncUntilComplete = debounceAsyncUntilComplete;
class TagUtils {
    /** Removes `oldTag` from the frontmatter tags of `note` and replaces it with `newTag`, if any. */
    static replaceTag({ note, oldTag, newTag, }) {
        if (lodash_1.default.isUndefined(note.tags) || lodash_1.default.isString(note.tags)) {
            note.tags = newTag;
        }
        else {
            const index = lodash_1.default.findIndex(note.tags, (tag) => tag === oldTag);
            if (newTag) {
                if (index >= 0) {
                    note.tags[index] = newTag;
                }
                else {
                    // Weird, can't find the old tag. Add the new one anyway.
                    note.tags.push(newTag);
                }
            }
            else {
                lodash_1.default.pull(note.tags, oldTag);
            }
        }
    }
}
exports.TagUtils = TagUtils;
class PublishUtils {
    static getSEOPropsFromConfig(config) {
        const { title, twitter, description: excerpt, image } = config.site;
        return {
            title,
            twitter,
            excerpt,
            image,
        };
    }
    static getSEOPropsFromNote(note) {
        const { title, created, updated, image } = note;
        const { excerpt, canonicalUrl, noindex, canonicalBaseUrl, twitter } = note.custom ? note.custom : {};
        return {
            title,
            excerpt,
            updated,
            created,
            canonicalBaseUrl,
            canonicalUrl,
            noindex,
            image,
            twitter,
        };
    }
}
exports.PublishUtils = PublishUtils;
class ConfigUtils {
    static genDefaultConfig() {
        const common = {
            useFMTitle: true,
            useNoteTitleForLink: true,
            mermaid: true,
            useKatex: true,
            dev: {
                enablePreviewV2: true,
            },
            site: {
                copyAssets: true,
                siteHierarchies: ["root"],
                siteRootDir: "docs",
                usePrettyRefs: true,
                title: "Dendron",
                description: "Personal knowledge space",
                siteLastModified: true,
                gh_edit_branch: "main",
            },
        };
        return {
            version: 4,
            ...common,
            commands: (0, intermediateConfigs_1.genDefaultCommandConfig)(),
            workspace: (0, intermediateConfigs_1.genDefaultWorkspaceConfig)(),
            preview: (0, intermediateConfigs_1.genDefaultPreviewConfig)(),
        };
    }
    // get
    static getProp(config, key) {
        const defaultConfig = ConfigUtils.genDefaultConfig();
        const configWithDefaults = lodash_1.default.defaultsDeep(config, defaultConfig);
        return configWithDefaults[key];
    }
    static getCommands(config) {
        return ConfigUtils.getProp(config, "commands");
    }
    static getWorkspace(config) {
        return ConfigUtils.getProp(config, "workspace");
    }
    static getPreview(config) {
        const out = ConfigUtils.getProp(config, "preview");
        // FIXME: for some reason, this can return undefined when run in context of chrome in `dendron-plugin-views`
        if (lodash_1.default.isUndefined(out)) {
            return ConfigUtils.genDefaultConfig().preview;
        }
        return out;
    }
    static getSite(config) {
        return ConfigUtils.getProp(config, "site");
    }
    static getVaults(config) {
        return ConfigUtils.getWorkspace(config).vaults;
    }
    static getHooks(config) {
        return ConfigUtils.getWorkspace(config).hooks;
    }
    static getJournal(config) {
        return ConfigUtils.getWorkspace(config).journal;
    }
    static getScratch(config) {
        return ConfigUtils.getWorkspace(config).scratch;
    }
    static getTask(config) {
        return ConfigUtils.getWorkspace(config).task;
    }
    static getLookup(config) {
        return ConfigUtils.getCommands(config).lookup;
    }
    static getEnableFMTitle(config, shouldApplyPublishRules) {
        return shouldApplyPublishRules
            ? ConfigUtils.getProp(config, "useFMTitle")
            : ConfigUtils.getPreview(config).enableFMTitle;
    }
    static getEnableNoteTitleForLink(config, shouldApplyPublishRules) {
        return shouldApplyPublishRules
            ? ConfigUtils.getProp(config, "useNoteTitleForLink")
            : ConfigUtils.getPreview(config).enableNoteTitleForLink;
    }
    static getEnableMermaid(config, shouldApplyPublishRules) {
        return shouldApplyPublishRules
            ? ConfigUtils.getProp(config, "mermaid")
            : ConfigUtils.getPreview(config).enableMermaid;
    }
    static getEnableKatex(config, shouldApplyPublishRules) {
        return shouldApplyPublishRules
            ? ConfigUtils.getProp(config, "useKatex")
            : ConfigUtils.getPreview(config).enableKatex;
    }
    static getSiteLogoUrl(config) {
        const { assetsPrefix, logo } = ConfigUtils.getSite(config);
        if (logo === undefined)
            return undefined;
        // Let's allow logos that are hosted off-site/in subdomains by passing in a full URL
        if ((0, regex_1.isWebUri)(logo))
            return logo;
        // Otherwise, this has to be an asset. It can't be anywhere else because of backwards compatibility.
        const logoBase = path_1.default.basename(logo); // Why just discard the rest of logo? Because that's what code used to do and I'm preserving backwards compatibility
        if (assetsPrefix) {
            const initialSlash = assetsPrefix.startsWith("/") ? "" : "/";
            return `${initialSlash}${assetsPrefix}/assets/${logoBase}`;
        }
        return `/assets/${logoBase}`;
    }
    static getEnablePrettyRefs(config, opts) {
        var _a, _b, _c;
        const override = (_c = (_b = (_a = opts === null || opts === void 0 ? void 0 : opts.note) === null || _a === void 0 ? void 0 : _a.config) === null || _b === void 0 ? void 0 : _b.global) === null || _c === void 0 ? void 0 : _c.enablePrettyRefs;
        if (override !== undefined)
            return override;
        return (opts === null || opts === void 0 ? void 0 : opts.shouldApplyPublishRules)
            ? ConfigUtils.getSite(config).usePrettyRefs
            : ConfigUtils.getPreview(config).enablePrettyRefs;
    }
    /**
     * NOTE: _config currently doesn't have a `global` object. We're keeping it here
     * to make using the API easier when we do add it
     */
    static getEnableChildLinks(_config, opts) {
        if (opts &&
            opts.note &&
            opts.note.config &&
            opts.note.config.global &&
            !lodash_1.default.isUndefined(opts.note.config.global.enableChildLinks)) {
            return opts.note.config.global.enableChildLinks;
        }
        return true;
    }
    // set
    static setProp(config, key, value) {
        lodash_1.default.set(config, key, value);
    }
    static setCommandsProp(config, key, value) {
        const path = `commands.${key}`;
        lodash_1.default.set(config, path, value);
    }
    static setWorkspaceProp(config, key, value) {
        const path = `workspace.${key}`;
        lodash_1.default.set(config, path, value);
    }
    static setVaults(config, value) {
        ConfigUtils.setWorkspaceProp(config, "vaults", value);
    }
    static setNoteLookupProps(config, key, value) {
        const path = `commands.lookup.note.${key}`;
        lodash_1.default.set(config, path, value);
    }
    static setJournalProps(config, key, value) {
        const path = `workspace.journal.${key}`;
        lodash_1.default.set(config, path, value);
    }
    static setScratchProps(config, key, value) {
        const path = `workspace.scratch.${key}`;
        lodash_1.default.set(config, path, value);
    }
    static setHooks(config, value) {
        ConfigUtils.setWorkspaceProp(config, "hooks", value);
    }
    static setPreviewProps(config, key, value) {
        const path = `preview.${key}`;
        lodash_1.default.set(config, path, value);
    }
    static configIsValid(opts) {
        const { clientVersion, configVersion } = opts;
        if (lodash_1.default.isUndefined(configVersion)) {
            throw new error_1.DendronError({
                message: "Cannot determine config version. Please make sure the field 'version' is present and correct",
                severity: constants_1.ERROR_SEVERITY.FATAL,
            });
        }
        const minCompatClientVersion = compat_1.CONFIG_TO_MINIMUM_COMPAT_MAPPING[configVersion];
        if (lodash_1.default.isUndefined(minCompatClientVersion)) {
            throw new error_1.DendronError({
                message: error_1.ErrorMessages.formatShouldNeverOccurMsg("Cannot find minimum compatible client version."),
                severity: constants_1.ERROR_SEVERITY.FATAL,
            });
        }
        const minCompatConfigVersion = lodash_1.default.findLastKey(compat_1.CONFIG_TO_MINIMUM_COMPAT_MAPPING, (ent) => {
            return semver_1.default.lte(ent, clientVersion);
        });
        if (lodash_1.default.isUndefined(minCompatConfigVersion)) {
            throw new error_1.DendronError({
                message: error_1.ErrorMessages.formatShouldNeverOccurMsg("cannot find minimum compatible config version."),
                severity: constants_1.ERROR_SEVERITY.FATAL,
            });
        }
        const clientVersionCompatible = semver_1.default.lte(minCompatClientVersion, clientVersion);
        const configVersionCompatible = Number(minCompatConfigVersion) <= configVersion;
        const isValid = clientVersionCompatible && configVersionCompatible;
        if (!isValid) {
            const reason = clientVersionCompatible ? "config" : "client";
            return {
                isValid,
                reason,
                minCompatClientVersion,
                minCompatConfigVersion,
            };
        }
        else {
            return { isValid, minCompatClientVersion, minCompatConfigVersion };
        }
    }
}
exports.ConfigUtils = ConfigUtils;
/**
 * Make name safe for dendron
 * @param name
 * @param opts
 */
function cleanName(name) {
    name = name
        .replace(new RegExp(lodash_1.default.escapeRegExp(path_1.default.sep), "g"), ".")
        .toLocaleLowerCase();
    name = name.replace(/ /g, "-");
    return name;
}
exports.cleanName = cleanName;
/** Given a path on any platform, convert it to a unix style path. Avoid using this with absolute paths. */
function normalizeUnixPath(fsPath) {
    return path_1.default.posix.format(path_1.default.parse(fsPath));
}
exports.normalizeUnixPath = normalizeUnixPath;
/** Wrapper(s) for easier testing, to wrap functions where we don't want to mock the global function. */
class Wrap {
    /** A useless wrapper around `setTimeout`. Useful for testing.
     *
     * If you are testing code that uses `setTimeout`, you can switch that code over to this wrapper instead,
     * and then mock the wrapper. We can't entirely mock `setTimeout` because that seems to break VSCode.
     */
    static setTimeout(callback, ms, ...args) {
        return setTimeout(callback, ms, ...args);
    }
}
exports.Wrap = Wrap;
